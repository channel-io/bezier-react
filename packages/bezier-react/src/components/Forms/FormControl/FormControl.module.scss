.FormLabelWrapper {
  &:where(.position-top) {
    margin-bottom: 4px;
    padding: 0 2px;
  }

  &:where(.position-left) {
    display: flex;
    grid-column: 1 / 1;
    grid-row: 1 / 1;
    align-items: center;
    align-self: start;

    height: var(--b-form-control-left-label-wrapper-height);
  }
}

.FormHelperTextWrapper {
  margin-top: 4px;
  padding: 0 2px;

  &:where(.position-left) {
    grid-column: 2;
    grid-row: 2 / 2;
  }
}

