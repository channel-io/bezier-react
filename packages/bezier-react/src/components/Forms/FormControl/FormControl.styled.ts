import { styled } from '~/src/foundation'

import { type InterpolationProps } from '~/src/types/Foundation'

const LEFT_LABEL_MIN_WIDTH = 150

const Box = styled.div<InterpolationProps>`
  position: relative;
  ${({ interpolation }) => interpolation}
`

export const Grid = styled(Box)`
  display: grid;
  grid-template-rows: repeat(2, auto);
  grid-template-columns: ${LEFT_LABEL_MIN_WIDTH}px 1fr;
  grid-column-gap: 12px;
  align-items: center;
`
