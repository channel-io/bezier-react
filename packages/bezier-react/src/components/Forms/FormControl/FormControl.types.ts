import type {
  AlphaBezierComponentProps,
  BezierComponentProps,
  ChildrenProps,
  IdentifierProps,
} from '~/src/types/ComponentProps'

import type {
  FormComponentProps,
  FormFieldSize,
} from '~/src/components/Forms'

type LabelPosition = 'top' | 'left'

interface FormControlOptions {
  leftLabelWrapperHeight?: FormFieldSize
  labelPosition?: LabelPosition
}

interface FormControlClassNameProps {
  className: string
}

export interface FormControlContextCommonValue extends Partial<IdentifierProps> {}

export interface FormControlAriaProps {
  'aria-labelledby'?: string
  'aria-describedby'?: string
}

interface CallbackRefProps {
  ref: (node: HTMLElement | null) => void
}

type PropsGetter<ExtraReturnType = {}> = <Props = {}>(props: Props) => Props & FormControlContextCommonValue & ExtraReturnType

export type GroupPropsGetter = PropsGetter<CallbackRefProps & FormControlAriaProps>

export type LabelPropsGetter = PropsGetter<FormControlClassNameProps>

export type FieldPropsGetter = PropsGetter<Omit<FormControlAriaProps, 'aria-labelledby'>>

export type HelperTextPropsGetter = PropsGetter<FormControlClassNameProps & CallbackRefProps & {
  visible: boolean
}>

export type ErrorMessagePropsGetter = HelperTextPropsGetter

export interface FormControlContextValue extends FormComponentProps {
  id: string
  labelId: string
  helperTextId: string
  errorMessageId: string
  getGroupProps: GroupPropsGetter
  getLabelProps: LabelPropsGetter
  getFieldProps: FieldPropsGetter
  getHelperTextProps: HelperTextPropsGetter
  getErrorMessageProps: ErrorMessagePropsGetter
}

export interface ContainerProps extends
  AlphaBezierComponentProps,
  ChildrenProps,
  Pick<FormControlOptions, 'labelPosition'> {}

export interface FormControlProps extends
  BezierComponentProps,
  ChildrenProps,
  FormComponentProps,
  FormControlOptions {}
