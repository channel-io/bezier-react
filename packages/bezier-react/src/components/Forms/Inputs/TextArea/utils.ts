import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
