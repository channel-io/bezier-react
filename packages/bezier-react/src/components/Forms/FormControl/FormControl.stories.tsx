/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import type { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import { SegmentedControl } from 'Components/Forms/SegmentedControl'
import { Radio } from 'Components/Forms/Radio'
import { Checkbox } from 'Components/Forms/Checkbox'
import { Text } from 'Components/Text'
import { FormGroup } from 'Components/Forms/FormGroup'
import { FormLabel } from 'Components/Forms/FormLabel'
import { TextField } from 'Components/Forms/Inputs/TextField'
import { TextArea } from 'Components/Forms/Inputs/TextArea'
import { Select } from 'Components/Forms/Inputs/Select'
import { FormHelperText, FormErrorMessage } from 'Components/Forms/FormHelperText'
import FormControl from './FormControl'
import FormControlProps from './FormControl.types'

export default {
  title: getTitle(base),
  component: FormControl,
} as Meta

const Template: Story<FormControlProps> = (args) => (
  <FormControl style={{ width: 400 }} {...args}>
    <FormLabel help="Lorem Ipsum">Label</FormLabel>
    <TextField placeholder="Placeholder" />
    <FormHelperText>Description</FormHelperText>
    <FormErrorMessage>Error!</FormErrorMessage>
  </FormControl>
)

export const Primary: Story<FormControlProps> = Template.bind({})
Primary.args = {
  id: 'form',
  labelPosition: 'top',
  leftLabelWrapperHeight: undefined,
  hasError: false,
  disabled: false,
  readOnly: false,
}

const WithMultiFormTemplate: Story<FormControlProps> = (args) => (
  <div style={{
    display: 'flex',
    flexDirection: 'column',
    gap: 16,
    width: 500,
  }}
  >
    <FormControl {...args}>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <TextField placeholder="Placeholder" />
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </FormControl>

    <FormControl {...args}>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <TextArea placeholder="Placeholder" />
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </FormControl>

    <FormControl {...args}>
      <FormLabel>Label</FormLabel>
      <Select />
    </FormControl>

    <FormControl {...args}>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <SegmentedControl>
        { ['Open', 'Snoozed', 'Closed'].map((item) => (
          <Text key={item} bold>{ item }</Text>
        )) }
      </SegmentedControl>
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </FormControl>

    <FormControl {...args}>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <FormGroup direction="row" gap={10}>
        <Checkbox>Option</Checkbox>
        <Checkbox>Option</Checkbox>
        <Checkbox>Option</Checkbox>
        <Checkbox>Option</Checkbox>
      </FormGroup>
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </FormControl>

    <FormControl {...args}>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <FormGroup direction="row" gap={20}>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            gap: 10,
          }}
        >
          <Radio>Immediately</Radio>
          <Radio>Year(s)</Radio>
          <Radio>Seasons</Radio>
        </div>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            gap: 10,
          }}
        >
          <Radio>Hour(s)</Radio>
          <Radio>Day(s)</Radio>
          <Radio>Minute(s)</Radio>
        </div>
      </FormGroup>
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </FormControl>
  </div>
)

export const WithMultiForm: Story<FormControlProps> = WithMultiFormTemplate.bind({})
WithMultiForm.args = {
  labelPosition: 'top',
  leftLabelWrapperHeight: undefined,
  hasError: false,
  disabled: false,
  readOnly: false,
}
