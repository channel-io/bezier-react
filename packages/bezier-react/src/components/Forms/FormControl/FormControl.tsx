/* External dependencies */
import React, { useState, useCallback, useMemo } from 'react'
import { isNil } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import useId from 'Hooks/useId'
import { omitBezierComponentProps, pickBezierComponentProps } from 'Utils/propsUtils'
import { TextFieldSize } from 'Components/Forms/Inputs/TextField'
import FormControlContext from './FormControlContext'
import FormControlProps, {
  GroupPropsGetter,
  FieldPropsGetter,
  LabelPropsGetter,
  HelperTextPropsGetter,
  ErrorMessagePropsGetter,
} from './FormControl.types'
import * as Styled from './FormControl.styled'

export const FORM_CONTROL_TEST_ID = 'bezier-react-form-control'

function FormControl({
  id: idProp,
  testId = FORM_CONTROL_TEST_ID,
  labelPosition = 'top',
  leftLabelWrapperHeight = TextFieldSize.M,
  children,
  ...rest
}: FormControlProps) {
  const [hasMultipleFields, setHasMultipleFields] = useState(false)
  const [hasHelperText, setHasHelperText] = useState(false)
  const [hasErrorMessage, setHasErrorMessage] = useState(false)

  const id = useId(idProp, 'field')
  const groupId = `${id}-group`
  const labelId = `${id}-label`
  const helperTextId = `${id}-help-text`
  const errorMessageId = `${id}-error-message`

  const fieldId = hasMultipleFields ? undefined : id

  const describerId = useMemo(() => {
    if (hasErrorMessage) { return errorMessageId }
    if (hasHelperText) { return helperTextId }
    return undefined
  }, [
    hasErrorMessage,
    hasHelperText,
    errorMessageId,
    helperTextId,
  ])

  const bezierProps = useMemo(() => pickBezierComponentProps(rest), [rest])
  const formCommonProps = useMemo(() => omitBezierComponentProps(rest), [rest])

  const getGroupProps = useCallback<GroupPropsGetter>(ownProps => ({
    id: groupId,
    'aria-labelledby': labelId,
    'aria-describedby': describerId,
    setIsRendered: setHasMultipleFields,
    ...ownProps,
  }), [
    groupId,
    labelId,
    describerId,
    setHasMultipleFields,
  ])

  const getLabelProps = useCallback<LabelPropsGetter>(ownProps => ({
    id: labelId,
    htmlFor: fieldId,
    typo: labelPosition === 'left' ? Typography.Size14 : Typography.Size13,
    Wrapper: labelPosition === 'top'
      ? Styled.TopLabelWrapper
      : (({ children: labelElement }) => (
        <Styled.LeftLabelWrapper height={leftLabelWrapperHeight}>
          { labelElement }
        </Styled.LeftLabelWrapper>
      )),
    ...ownProps,
  }), [
    fieldId,
    labelId,
    labelPosition,
    leftLabelWrapperHeight,
  ])

  const getFieldProps = useCallback<FieldPropsGetter>(ownProps => ({
    id: fieldId,
    'aria-describedby': hasMultipleFields ? undefined : describerId,
    ...formCommonProps,
    ...ownProps,
  }), [
    fieldId,
    describerId,
    formCommonProps,
    hasMultipleFields,
  ])

  const getHelperTextProps = useCallback<HelperTextPropsGetter>(ownProps => ({
    id: helperTextId,
    visible: isNil(formCommonProps?.hasError) || !formCommonProps?.hasError,
    setIsRendered: setHasHelperText,
    Wrapper: labelPosition === 'top'
      ? Styled.TopHelperTextWrapper
      : Styled.LeftHelperTextWrapper,
    ...ownProps,
  }), [
    helperTextId,
    labelPosition,
    formCommonProps,
  ])

  const getErrorMessageProps = useCallback<ErrorMessagePropsGetter>(ownProps => ({
    id: errorMessageId,
    visible: isNil(formCommonProps?.hasError) || formCommonProps?.hasError,
    setIsRendered: setHasErrorMessage,
    Wrapper: labelPosition === 'top'
      ? Styled.TopHelperTextWrapper
      : Styled.LeftHelperTextWrapper,
    ...ownProps,
  }), [
    errorMessageId,
    labelPosition,
    formCommonProps,
  ])

  const contextValue = useMemo(() => ({
    id,
    labelId,
    helperTextId,
    errorMessageId,
    getGroupProps,
    getLabelProps,
    getFieldProps,
    getHelperTextProps,
    getErrorMessageProps,
    ...formCommonProps,
  }), [
    id,
    labelId,
    helperTextId,
    errorMessageId,
    getGroupProps,
    getLabelProps,
    getFieldProps,
    getHelperTextProps,
    getErrorMessageProps,
    formCommonProps,
  ])

  if (!children) { return null }

  const Container = {
    top: Styled.Box,
    left: Styled.Grid,
  }[labelPosition]

  return (
    <FormControlContext.Provider value={contextValue}>
      <Container
        data-testid={testId}
        {...bezierProps}
      >
        { children }
      </Container>
    </FormControlContext.Provider>
  )
}

export default FormControl
