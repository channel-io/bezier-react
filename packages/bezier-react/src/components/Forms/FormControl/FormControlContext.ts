/* External dependencies */
import { createContext } from 'react'

/* Internal dependencies */
import { FormControlContextValue } from '~/src/components/Forms/FormControl'

const FormControlContext = createContext<FormControlContextValue | undefined>(undefined)

export default FormControlContext
