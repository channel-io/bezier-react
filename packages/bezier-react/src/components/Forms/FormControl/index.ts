import FormControl from './FormControl'
import FormControlContext from './FormControlContext'
import type FormControlProps from './FormControl.types'
import type {
  FormControlContextValue,
  FormControlContextCommonValue,
  FormControlAriaProps,
} from './FormControl.types'

export type {
  FormControlProps,
  FormControlContextValue,
  FormControlContextCommonValue,
  FormControlAriaProps,
}

export {
  FormControl,
  FormControlContext,
}
