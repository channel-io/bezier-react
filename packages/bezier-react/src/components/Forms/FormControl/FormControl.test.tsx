/* External dependencies */
import React from 'react'
import { isInaccessible } from '@testing-library/react'

/* Internal dependencies */
import { render } from '~/src/utils/testUtils'
import {
  SingleFieldForm,
  MultipleFieldForm,
  SingleFieldFormWithLabelFont,
  MOCK_CONSTS,
} from './__mocks__/forms'
import FormControl, { FORM_CONTROL_TEST_ID } from './FormControl'
import type FormControlProps from './FormControl.types'

describe('FormControl >', () => {
  let props: FormControlProps

  beforeEach(() => {
    props = {
      id: 'form',
      labelPosition: 'top',
      leftLabelWrapperHeight: undefined,
      hasError: false,
      disabled: false,
      readOnly: false,
    }
  })

  const renderComponent = ({ children, ...rest }: FormControlProps) => render(
    <FormControl {...props} {...rest}>
      { children }
    </FormControl>,
  )

  describe('Snapshot >', () => {
    it('With single field', () => {
      const { getByTestId } = renderComponent({
        children: SingleFieldForm,
      })

      const rendered = getByTestId(FORM_CONTROL_TEST_ID)
      expect(rendered).toMatchSnapshot()
    })

    it('With single field and left label position', () => {
      const { getByTestId } = renderComponent({
        labelPosition: 'left',
        children: SingleFieldForm,
      })

      const rendered = getByTestId(FORM_CONTROL_TEST_ID)
      expect(rendered).toMatchSnapshot()
    })

    it('With multiple field', () => {
      const { container } = renderComponent({
        children: MultipleFieldForm,
      })

      expect(container).toMatchSnapshot()
    })

    it('With multiple field and left label position', () => {
      const { container } = renderComponent({
        labelPosition: 'left',
        children: MultipleFieldForm,
      })

      expect(container).toMatchSnapshot()
    })
  })

  describe('Accessibility >', () => {
    it('With single field', () => {
      const { container } = renderComponent({
        children: SingleFieldForm,
      })

      expect(isInaccessible(container)).toBe(false)
    })

    it('With multiple field', () => {
      const { container } = renderComponent({
        children: MultipleFieldForm,
      })

      expect(isInaccessible(container)).toBe(false)
    })
  })

  describe('Single field form >', () => {
    it('The FormLabel component should have the correct attribute', () => {
      const { getByText, getByLabelText } = renderComponent({
        children: SingleFieldForm,
      })

      const labelNode = getByText(MOCK_CONSTS.LABEL_TEXT)
      const inputId = getByLabelText(MOCK_CONSTS.LABEL_TEXT).id

      expect(labelNode).toHaveAttribute('for', inputId)
      expect(labelNode.id).toBe(`${inputId}-label`)
    })

    it('The FormLabel component should have the correct style it depends on top label position', () => {
      const { getByText } = renderComponent({
        children: SingleFieldForm,
      })

      const labelNode = getByText(MOCK_CONSTS.LABEL_TEXT)

      expect(labelNode).toHaveStyle({
        'font-size': '1.3rem',
      })
    })

    it('The FormLabel component should have the correct style it depends on left label position', () => {
      const { getByText } = renderComponent({
        labelPosition: 'left',
        children: SingleFieldForm,
      })

      const labelNode = getByText(MOCK_CONSTS.LABEL_TEXT)

      expect(labelNode).toHaveStyle({
        'font-size': '1.4rem',
      })
    })

    it('The FormLabel component should have the correct style with own typo props', () => {
      const { getByText } = renderComponent({
        labelPosition: 'left',
        children: SingleFieldFormWithLabelFont,
      })

      const labelNode = getByText(MOCK_CONSTS.LABEL_TEXT)

      expect(labelNode).toHaveStyle({
        'font-size': '1.8rem',
      })
    })

    it('The Field(Input) component should have the correct attribute', () => {
      const { getByText, getByLabelText } = renderComponent({
        children: SingleFieldForm,
      })

      const inputNode = getByLabelText(MOCK_CONSTS.LABEL_TEXT)
      const helperTextNode = getByText(MOCK_CONSTS.HELPER_TEXT_TEXT)

      expect(inputNode).toHaveAttribute('aria-describedby', helperTextNode.id)
    })

    it('The Field(Input) component should have the correct attribute when the "hasError" prop is true', () => {
      const { getByText, getByLabelText } = renderComponent({
        children: SingleFieldForm,
        hasError: true,
      })

      const inputNode = getByLabelText(MOCK_CONSTS.LABEL_TEXT)
      const errorMessageNode = getByText(MOCK_CONSTS.ERROR_MESSAGE_TEXT)

      expect(inputNode).toHaveAttribute('aria-describedby', errorMessageNode.id)
    })

    it('The FormHelperText component is only visible when the "hasError" prop is false', () => {
      const { getByText, queryByText } = renderComponent({
        children: SingleFieldForm,
        hasError: false,
      })

      const helperTextNode = getByText(MOCK_CONSTS.HELPER_TEXT_TEXT)
      const errorMessageNode = queryByText(MOCK_CONSTS.ERROR_MESSAGE_TEXT)

      expect(helperTextNode).toBeVisible()
      expect(errorMessageNode).toBeNull()
    })

    it('The FormErrorMessage component is only visible when the "hasError" prop is true', () => {
      const { getByText, queryByText } = renderComponent({
        children: SingleFieldForm,
        hasError: true,
      })

      const errorMessageNode = getByText(MOCK_CONSTS.ERROR_MESSAGE_TEXT)
      const helperTextNode = queryByText(MOCK_CONSTS.HELPER_TEXT_TEXT)

      expect(errorMessageNode).toBeVisible()
      expect(helperTextNode).toBeNull()
    })
  })

  describe('Multiple(Group) field form >', () => {
    it('The FormLabel for group element should not have "for" attribute', () => {
      const { getByText } = renderComponent({
        children: MultipleFieldForm,
      })

      const groupLabelNode = getByText(MOCK_CONSTS.LABEL_TEXT)

      expect(groupLabelNode).not.toHaveAttribute('for')
    })

    it('The FormGroup component should have the correct attribute', () => {
      const { getByRole, getByText } = renderComponent({
        children: MultipleFieldForm,
      })

      const groupNode = getByRole('group')
      const groupLabelNode = getByText(MOCK_CONSTS.LABEL_TEXT)
      const groupHelperTextNode = getByText(MOCK_CONSTS.HELPER_TEXT_TEXT)

      expect(groupNode).toHaveAttribute('aria-labelledby', groupLabelNode.id)
      expect(groupNode).toHaveAttribute('aria-describedby', groupHelperTextNode.id)
    })

    it('The FormGroup component should have the correct attribute when the "hasError" prop is true', () => {
      const { getByRole, getByText } = renderComponent({
        children: MultipleFieldForm,
        hasError: true,
      })

      const groupNode = getByRole('group')
      const groupLabelNode = getByText(MOCK_CONSTS.LABEL_TEXT)
      const groupErrorMessageNode = getByText(MOCK_CONSTS.ERROR_MESSAGE_TEXT)

      expect(groupNode).toHaveAttribute('aria-labelledby', groupLabelNode.id)
      expect(groupNode).toHaveAttribute('aria-describedby', groupErrorMessageNode.id)
    })

    it('The Field(Input) component inside FormGroup should have a unique ID that is not FormGroup ID', () => {
      const { getByRole, getByLabelText } = renderComponent({
        children: MultipleFieldForm,
      })

      const groupNodeId = getByRole('group').id
      const firstInnerInputNodeId = getByLabelText(MOCK_CONSTS.FIRST_FIELD_LABEL_TEXT).id
      const secondInnerInputNodeId = getByLabelText(MOCK_CONSTS.SECOND_FIELD_LABEL_TEXT).id

      expect(firstInnerInputNodeId).not.toEqual(groupNodeId)
      expect(secondInnerInputNodeId).not.toEqual(groupNodeId)
      expect(firstInnerInputNodeId).not.toEqual(secondInnerInputNodeId)
    })
  })
})
