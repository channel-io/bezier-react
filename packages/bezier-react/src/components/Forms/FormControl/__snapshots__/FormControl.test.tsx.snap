// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl > Snapshot > With multiple field 1`] = `
<div>
  <div
    class="Stack display-flex direction-vertical margin layout"
    data-testid="bezier-react-form-control"
    style="--b-form-control-left-label-wrapper-height: 36px;"
  >
    <label
      class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
      data-testid="bezier-react-form-label"
      id="form-label"
      style="--b-text-color: var(--txt-black-darkest);"
    >
      Label
    </label>
    <div
      aria-describedby="form-help-text"
      aria-labelledby="form-label"
      class="Stack display-flex direction-vertical justify-start align-stretch wrap margin layout"
      data-testid="bezier-react-form-group"
      id="form-group"
      role="group"
      style="--b-stack-spacing: 6px;"
    >
      <div
        class="Stack display-flex direction-vertical margin layout"
        data-testid="bezier-react-form-control"
        style="--b-form-control-left-label-wrapper-height: 36px;"
      >
        <label
          class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
          data-testid="bezier-react-form-label"
          for="field-:r3:"
          id="field-:r3:-label"
          style="--b-text-color: var(--txt-black-darkest);"
        >
          First Inner Label
        </label>
        <div
          class="TextFieldWrapper variant-primary size-36"
          data-testid="bezier-react-text-input"
        >
          <input
            autocomplete="off"
            class="TextFieldInput"
            id="field-:r3:"
            size="36"
            value=""
          />
        </div>
      </div>
      <div
        class="Stack display-flex direction-vertical margin layout"
        data-testid="bezier-react-form-control"
        style="--b-form-control-left-label-wrapper-height: 36px;"
      >
        <label
          class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
          data-testid="bezier-react-form-label"
          for="field-:r4:"
          id="field-:r4:-label"
          style="--b-text-color: var(--txt-black-darkest);"
        >
          Second Inner Label
        </label>
        <div
          class="TextFieldWrapper variant-primary size-36"
          data-testid="bezier-react-text-input"
        >
          <input
            aria-invalid="true"
            autocomplete="off"
            class="TextFieldInput"
            id="field-:r4:"
            size="36"
            value=""
          />
        </div>
      </div>
    </div>
    <p
      class="Text typo-13 align-left margin FormHelperText FormHelperTextWrapper"
      data-testid="bezier-react-form-helper-text"
      id="form-help-text"
      style="--b-text-color: var(--txt-black-dark);"
    >
      Description
    </p>
  </div>
</div>
`;

exports[`FormControl > Snapshot > With multiple field and left label position 1`] = `
.c0 {
  position: relative;
}

.c1 {
  display: grid;
  grid-template-rows: repeat(2,auto);
  grid-template-columns: 150px 1fr;
  grid-column-gap: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <div
    class="c0 c1"
    data-testid="bezier-react-form-control"
    style="--b-form-control-left-label-wrapper-height: 36px;"
  >
    <label
      class="Text typo-14 bold align-left margin LabelText FormLabelWrapper position-left"
      data-testid="bezier-react-form-label"
      id="form-label"
      style="--b-text-color: var(--txt-black-darkest);"
    >
      Label
    </label>
    <div
      aria-describedby="form-help-text"
      aria-labelledby="form-label"
      class="Stack display-flex direction-vertical justify-start align-stretch wrap margin layout"
      data-testid="bezier-react-form-group"
      id="form-group"
      role="group"
      style="--b-stack-spacing: 6px;"
    >
      <div
        class="Stack display-flex direction-vertical margin layout"
        data-testid="bezier-react-form-control"
        style="--b-form-control-left-label-wrapper-height: 36px;"
      >
        <label
          class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
          data-testid="bezier-react-form-label"
          for="field-:r6:"
          id="field-:r6:-label"
          style="--b-text-color: var(--txt-black-darkest);"
        >
          First Inner Label
        </label>
        <div
          class="TextFieldWrapper variant-primary size-36"
          data-testid="bezier-react-text-input"
        >
          <input
            autocomplete="off"
            class="TextFieldInput"
            id="field-:r6:"
            size="36"
            value=""
          />
        </div>
      </div>
      <div
        class="Stack display-flex direction-vertical margin layout"
        data-testid="bezier-react-form-control"
        style="--b-form-control-left-label-wrapper-height: 36px;"
      >
        <label
          class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
          data-testid="bezier-react-form-label"
          for="field-:r7:"
          id="field-:r7:-label"
          style="--b-text-color: var(--txt-black-darkest);"
        >
          Second Inner Label
        </label>
        <div
          class="TextFieldWrapper variant-primary size-36"
          data-testid="bezier-react-text-input"
        >
          <input
            aria-invalid="true"
            autocomplete="off"
            class="TextFieldInput"
            id="field-:r7:"
            size="36"
            value=""
          />
        </div>
      </div>
    </div>
    <p
      class="Text typo-13 align-left margin FormHelperText FormHelperTextWrapper position-left"
      data-testid="bezier-react-form-helper-text"
      id="form-help-text"
      style="--b-text-color: var(--txt-black-dark);"
    >
      Description
    </p>
  </div>
</div>
`;

exports[`FormControl > Snapshot > With single field 1`] = `
<div
  class="Stack display-flex direction-vertical margin layout"
  data-testid="bezier-react-form-control"
  style="--b-form-control-left-label-wrapper-height: 36px;"
>
  <label
    class="Text typo-13 bold align-left margin LabelText FormLabelWrapper position-top"
    data-testid="bezier-react-form-label"
    for="form"
    id="form-label"
    style="--b-text-color: var(--txt-black-darkest);"
  >
    Label
  </label>
  <div
    class="TextFieldWrapper variant-primary size-36"
    data-testid="bezier-react-text-input"
  >
    <input
      aria-describedby="form-help-text"
      autocomplete="off"
      class="TextFieldInput"
      id="form"
      size="36"
      value=""
    />
  </div>
  <p
    class="Text typo-13 align-left margin FormHelperText FormHelperTextWrapper"
    data-testid="bezier-react-form-helper-text"
    id="form-help-text"
    style="--b-text-color: var(--txt-black-dark);"
  >
    Description
  </p>
</div>
`;

exports[`FormControl > Snapshot > With single field and left label position 1`] = `
.c0 {
  position: relative;
}

.c1 {
  display: grid;
  grid-template-rows: repeat(2,auto);
  grid-template-columns: 150px 1fr;
  grid-column-gap: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  class="c0 c1"
  data-testid="bezier-react-form-control"
  style="--b-form-control-left-label-wrapper-height: 36px;"
>
  <label
    class="Text typo-14 bold align-left margin LabelText FormLabelWrapper position-left"
    data-testid="bezier-react-form-label"
    for="form"
    id="form-label"
    style="--b-text-color: var(--txt-black-darkest);"
  >
    Label
  </label>
  <div
    class="TextFieldWrapper variant-primary size-36"
    data-testid="bezier-react-text-input"
  >
    <input
      aria-describedby="form-help-text"
      autocomplete="off"
      class="TextFieldInput"
      id="form"
      size="36"
      value=""
    />
  </div>
  <p
    class="Text typo-13 align-left margin FormHelperText FormHelperTextWrapper position-left"
    data-testid="bezier-react-form-helper-text"
    id="form-help-text"
    style="--b-text-color: var(--txt-black-dark);"
  >
    Description
  </p>
</div>
`;
